crates/ruff_linter/src/rules/ruff/rules/sort_dunder_all.rs

```diff
@@ -0,0 +1,499 @@
+                }
+                parentheses_open = true;
+            }
+            Tok::Rpar | Tok::Rsqb | Tok::Newline => {
+                if !(items_in_line.is_empty() && comment_in_line.is_none()) {
+                    lines.push(DunderAllLine::new(
+                        &items_in_line,
+                        comment_in_line,
+                        range.start(),
+                    ));
+                }
+                break;
+            }
+            Tok::NonLogicalNewline => {
+                if !(items_in_line.is_empty() && comment_in_line.is_none()) {
+                    lines.push(DunderAllLine::new(
+                        &items_in_line,
+                        comment_in_line,
+                        range.start(),
+                    ));
+                    items_in_line.clear();
+                    comment_in_line = None;
+                }
+            }
+            Tok::Comment(_) => comment_in_line = Some(subrange),
+            Tok::String { value, .. } => items_in_line.push((value, subrange)),
+            Tok::Comma => continue,
+            _ => return None,
+        }
+    }
+    Some(lines)
+}
+
+#[derive(Debug)]
+struct DunderAllLine {
+    items: Vec<(String, TextRange)>,
+    comment: Option<TextRange>,
+}
+
+impl DunderAllLine {
+    fn new(items: &[(String, TextRange)], comment: Option<TextRange>, offset: TextSize) -> Self {
+        assert!(comment.is_some() || !items.is_empty());
+        Self {
+            items: items
+                .iter()
+                .map(|(s, r)| (s.to_owned(), r + offset))
+                .collect(),
+            comment: comment.map(|c| c + offset),
+        }
+    }
+}
+
+fn collect_dunder_all_items(lines: &[DunderAllLine]) -> Vec<DunderAllItem> {
+    // Given data on each line in `__all__`, group lines together into "items".
+    // Each item contains exactly one element,
+    // but might contain multiple comments attached to that element
+    // that must move with the element when `__all__` is sorted.
+    let mut all_items = vec![];
+    let mut this_range = None;
+    let mut idx = 0;
+    for line in lines {
+        let DunderAllLine { items, comment } = line;
+        match (items.as_slice(), comment) {
+            ([], Some(_)) => {
+                this_range = *comment;
+            }
+            ([(first_val, first_range), rest @ ..], _) => {
+                let range = this_range.map_or(*first_range, |r| {
+                    TextRange::new(r.start(), first_range.end())
+                });
+                all_items.push(DunderAllItem {
+                    value: first_val.clone(),
+                    original_index: idx,
+                    range,
+                    additional_comments: *comment,
+                });
+                this_range = None;
+                idx += 1;
+                for (value, range) in rest {
+                    all_items.push(DunderAllItem {
+                        value: value.clone(),
```

Comment:
I feel like it should be possible to avoid this clone, since we don't need `lines` beyond this method. We could consider either making `lines` an owned `Vec` to this method, or using lifetimes on `DunderAllItem`, and having it take `&str` that's tied to the lifetime of `lines: &[DunderAllLine]`. (I'd probably do the latter.)

Reply:
I think I initially tried doing the latter, but the borrow checker was very angry with me, so I ran away. I'll have another go tomorrow.